# HW07 Week 10 Multiple linear regression

### Question 1
1. the difference between Simple Linear Regression and Multiple Linear Regression; and the benefit the latter provides over the former
Simple linear regression is model of the relationship between one independent and one dependent variable and predict the outcome based on the change in predictor and fitted straight line. The multiple linear regression models relationship between several independent variable and only one dependent which this allows us to have a better understanding of factors indluencing the outcome. The good thing about the it is that it can account multiple factors simultaneously, gives a more accurate prediction of and understanding of how factors influences the outcome.  

2. the difference between using a continuous variable and an indicator variable in Simple Linear Regression; and these two linear forms
Using a continous variable is basically expressing numerical data, we are able to to take any value within the given range, and our regression line will follow along the spectrum of values. An indicator variable presents as categorical data, we would have to convert them into parameters that code can express them.

3. the change that happens in the behavior of the model (i.e., the expected nature of the data it models) when a single indicator variable is introduced alongside a continuous variable to create a Multiple Linear Regression; and these two linear forms (i.e., the Simple Linear Regression versus the Multiple Linear Regression)
When adding a indicator variable along with thew continous variable, the model adjust its response to continous variable basaed on the indicators presence or absense. Meaning where will be two different regression line, which predicts outcomes differently for each category. 

4. the effect of adding an interaction between a continuous and an indicator variable in Multiple Linear Regression models; and this linear form
The interaction between the two variables helps them to produce a more flexible fit in which that the relationship between them differs based on how they are presented. 

5. the behavior of a Multiple Linear Regression model (i.e., the expected nature of the data it models) based only on indicator variables derived from a non-binary categorical variable; this linear form; and the necessarily resulting binary variable encodings it utilizes
A non-binary categorical variable presents using only indicator variable would cause each unique category would be encoded and seperate as binary variables, which lets the model to handle multiple categories. And enable the model to create distinct intercepts for each of the categories and adjust the predicted outcome. 

### Question 2
Explain in your own words (but working with a ChatBot if needed) what the specific (outcome and predictor) variables are for the scenario below; whether or not any meaningful interactions might need to be taken into account when predicting the outcome; and provide the linear forms with and without the potential interactions that might need to be considered
Imagine a company that sells sports equipment. The company runs advertising campaigns on TV and online platforms. The effectiveness of the TV ad might depend on the amount spent on online advertising and vice versa, leading to an interaction effect between the two advertising mediums.

1. Explain how to use these two formulas to make predictions of the outcome, and give a high level explaination in general terms of the difference between predictions from the models with and without the interaction

$ Sales = β_0 + β_1 (TV Advertising) + β_2(online advertising )$ is the equation with the interaction. We simply just adds up the effects of each ads seperately to give a sales prdiction.  $Sales=β_0 + β_1 (TV Advertising) + β_2(Online Advertising) + β_3 (TV Advertising×Online Advertising)$ is the equation with include the interaction within the model. By using this formula, we includes interaction which adjust the sale prediction under the consideration of how TV and online ads might change each other's impact. 

2. Explain how to update and use the implied two formulas to make predictions of the outcome if, rather than considering two continuous predictor variables, we instead suppose the advertisement budgets are simply categorized as either "high" or "low" (binary variables)

As we excludes interaction in this model, it predicts sales by simply adding the separate effects of high TV and high online advertising budgets. Each type of advertising affects sales independently, meaning the impact of one does not depend on the other. When we include interaction in this model, it includes an additional term that captures a combined effect when both advertising budgets are high. This interaction term allows for a potentially stronger or different sales impact when both types of ad spending are high, rather than treating them as completely separate effects.

### Question 3
Would you please explain this in class, i really don't understand what to start here. 

### Question 4
Explain the apparent contradiction between the factual statements regarding the fit below that "the model only explains 17.6% of the variability in the data" while at the same time "many of the coefficients are larger than 10 while having strong or very strong evidence against the null hypothesis of 'no effect'"

The R-squared of 17.6% indicates that the model explains only a small portion of the total variability in HP, meaning most of the variability is due to factors not included in the model. However, the significant p-values for many coefficients (with evidence strongly against the null hypothesis) suggest that individual predictors, like Generation and Sp. Def, have a meaningful relationship with HP. This means that while these predictors are statistically significant and show large effects (coefficients greater than 10), they still do not explain a large portion of the variability in the outcome, likely because there are other important factors not captured by the model. Therefore, these two results are not as differ but reflect the fact that the predictors can have significant effects while the model may not account for most of the variation.


### Question 5
 Discuss the following (five cells of) code and results with a ChatBot and based on the understanding you arrive at in this conversation explain what the following (five cells of) are illustrating


In [22]:
#Cell 1
import numpy as np
from sklearn.model_selection import train_test_split

fifty_fifty_split_size = int(pokeaman.shape[0]*0.5)

# Replace "NaN" (in the "Type 2" column with "None")
pokeaman.fillna('None', inplace=True)

np.random.seed(130)
pokeaman_train,pokeaman_test = \
  train_test_split(pokeaman, train_size=fifty_fifty_split_size)
pokeaman_train

#                     Name   Type 1   Type 2  HP  Attack  Defense  \
370  338                  Solrock     Rock  Psychic  70      95       85   
6      6                Charizard     Fire   Flying  78      84       78   
242  224                Octillery    Water     None  75     105       75   
661  600                    Klang    Steel     None  60      80       95   
288  265                  Wurmple      Bug     None  45      45       35   
..   ...                      ...      ...      ...  ..     ...      ...   
522  471                  Glaceon      Ice     None  65      60      110   
243  225                 Delibird      Ice   Flying  45      55       45   
797  720      HoopaHoopa Confined  Psychic    Ghost  80     110       60   
117  109                  Koffing   Poison     None  40      65       95   
409  373  SalamenceMega Salamence   Dragon   Flying  95     145      130   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
370       55       65     70           3      False  
6        109       85    100           1      False  
242      105       75     45           2      False  
661       70       85     50           5      False  
288       20       30     20           3      False  
..       ...      ...    ...         ...        ...  
522      130       95     65           4      False  
243       65       45     75           2      False  
797      150      130     70           6       True  
117       60       45     35           1      False  
409      120       90    120           3      False  

[400 rows x 12 columns]

In cell one, after imported the nesseary libaries, the data set split into 50% for training and 50% for testing. 

In [23]:
# cell 2
model_spec3 = smf.ols(formula='HP ~ Attack + Defense', 
                      data=pokeaman_train)
model3_fit = model_spec3.fit()
model3_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        03:48:02   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     42.5882      3.580     11.897      0.000      35.551      49.626
Attack         0.2472      0.041      6.051      0.000       0.167       0.327
Defense        0.1001      0.045      2.201      0.028       0.011       0.190
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         343.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Cell 2 uses smf.ols method to specify an OLS regression model, with the formula of HP ~ Attack + Defense indicating that the target variable Hp is being modeled as a linear combination of Attack and Defense variables.

In [24]:
#Cell 3
yhat_model3 = model3_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model3_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model3)[0,1]**2)

'In sample' R-squared:     0.14771558304519894
'Out of sample' R-squared: 0.21208501873920738


The thrid cell uses the predict method to generate prediction for HP based on pokeman_test using fitted model.Then model3_fit.rsquared denmonstrates how the model explains the variation in HP variable in the training data. 

In [25]:
#Cell 4
model4_linear_form = 'HP ~ Attack * Defense * Speed * Legendary'
model4_linear_form += ' * Q("Sp. Def") * Q("Sp. Atk")'
# DO NOT try adding '* C(Generation) * C(Q("Type 1")) * C(Q("Type 2"))'
# That's 6*18*19 = 6*18*19 possible interaction combinations...
# ...a huge number that will blow up your computer

model4_spec = smf.ols(formula=model4_linear_form, data=pokeaman_train)
model4_fit = model4_spec.fit()
model4_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.467
Model:                            OLS   Adj. R-squared:                  0.369
Method:                 Least Squares   F-statistic:                     4.764
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           4.23e-21
Time:                        03:48:02   Log-Likelihood:                -1738.6
No. Observations:                 400   AIC:                             3603.
Df Residuals:                     337   BIC:                             3855.
Df Model:                          62                                         
Covariance Type:            nonrobust                                         
====================================================================================================================================
                                                                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------------
Intercept                                                          521.5715    130.273      4.004      0.000     265.322     777.821
Legendary[T.True]                                                   -6.1179      2.846     -2.150      0.032     -11.716      -0.520
Attack                                                              -8.1938      2.329     -3.518      0.000     -12.775      -3.612
Attack:Legendary[T.True]                                         -1224.9610    545.105     -2.247      0.025   -2297.199    -152.723
Defense                                                             -6.1989      2.174     -2.851      0.005     -10.475      -1.923
Defense:Legendary[T.True]                                         -102.4030     96.565     -1.060      0.290    -292.350      87.544
Attack:Defense                                                       0.0985      0.033      2.982      0.003       0.034       0.164
Attack:Defense:Legendary[T.True]                                    14.6361      6.267      2.336      0.020       2.310      26.963
Speed                                                               -7.2261      2.178     -3.318      0.001     -11.511      -2.942
Speed:Legendary[T.True]                                            704.8798    337.855      2.086      0.038      40.309    1369.450
Attack:Speed                                                         0.1264      0.038      3.351      0.001       0.052       0.201
Attack:Speed:Legendary[T.True]                                       5.8648      2.692      2.179      0.030       0.570      11.160
Defense:Speed                                                        0.1026      0.039      2.634      0.009       0.026       0.179
Defense:Speed:Legendary[T.True]                                     -6.9266      3.465     -1.999      0.046     -13.742      -0.111
Attack:Defense:Speed                                                -0.0016      0.001     -2.837      0.005      -0.003      -0.001
Attack:Defense:Speed:Legendary[T.True]                              -0.0743      0.030     -2.477      0.014      -0.133      -0.015
Q("Sp. Def")                                                        -5.3982      1.938     -2.785      0.006      -9.211      -1.586
Legendary[T.True]:Q("Sp. Def")                                    -282.2496    126.835     -2.225      0.027    -531.738     -32.761
Attack:Q("Sp. Def")                                                  0.1094      0.034      3.233      0.001       0.043       0.176
Attack:Legendary[T.True]:Q("Sp. Def")                               12.6503      5.851      2.162      0.031       1.141      24.160
Defense:Q("Sp. Def")    

In the fourth cell, it denmonstrates a more complex model specification using the interaction variables with a formula of HP ~ Attack * Defense * Speed * Legendary. The Q() uses to handle certain categorical variable such as sp.def and sp.atk which need to use them in a different way. Then the using the model4_linear_form, we formed a more complex fitted OLS model with the .summary() method provides the fit result.  

In [26]:
#Cell 5
yhat_model4 = model4_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model4_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model4)[0,1]**2)

'In sample' R-squared:     0.46709442115833855
'Out of sample' R-squared: 0.002485342598992873


In cell 5, the code yhat_model4 demonstrates the prediction made for the test and the R-squared values are computed. 

### Question 6
Basically when predictors are stronly correlated, it is very difficult for the model to isolate the effect of each individual predictors, which gives standard error for the coefficients so that our prediction are no long reliable. In model 14, the model starts with the variable then created interaction terms which combines some of these variable together as a whole. Instead of lookind how each one of the interact the prediction, it checks how they together changes the outcome. The conditional number we uses tells use how multicollinearity behaves in the model, and model4 denmonstrates a high value which shows that it is not reliable. When the model does a well job fitting training data but not new data, this is where overfitting comes in. Compare to model4, model3, is not as complex since it doesn't haven as many predictors and interactions. This makes it less likely to cause overfitting and multicollinearity, thus it will then be more accurate when making prediction on new data. 

### Question 7
Model5 uses the same preictors that model3 uses, but then adds one more variable that could possibly explain the variation in HP. Specifically sp.Atk and categorical variables like generation, Types of pokemon. Then this model removes omse of the interaction terms to focus only on the main effect that the model looks at each variable's individual contribution to prediction of HP, which is an improvement from model4 being too complex. 

Model6 uses as an extension from model5, where since model5 took some new variables to look for some hidden affect that thoise variable may impact, model6 removes the variables that were not contributing and espacially focuses on the ones were. Then it adds in interaction and more combination of some variables that affects the HP. As an example, it compares the generations with certain type 1 categories such as normal and water. 

For model7, it takes the variables in model6 and adds in more interaction, mostly betweet variables attack, speed, sp.def and sp.atk. Using the interaction of those variables, this model capture more complex relationships that potentially affects HP. It improves model's ability to capture the true pattern in thge data by understanding how combination of multiple factor work together can do proper predictions. However, i does keeps the binary indicator which is very helpful for it to account for different categories. Thus it made an extension of being more accurate then model6. 

### Question 9
Throught the given R values we recieved, it denmonstrates that model6 for in-sample R-squared value from range 0.33 to 0.44 does a good job in targeting variable. The out of sample R-squared is consistenly lower then the in-sample but is still moderate with values being around 0.23 and 0.29, however, this drop does suggests a sign for overfitting, as comparing to model7's, it indicates that model6 generalizes relatively well to new data. Meanwhile model6 perform very consistent through different generations, having more stable out of sample R squared values compare to model7. For model7, it does a better job ranging from 0.37 to 0.57, denmonstrating it is better when it comes to fitting the training data. 

### Chatbot session 
Here’s a summary of the conversation, mentioning Models 1 to 7:

We discussed a series of regression models (Models 1 to 7) to predict a target variable. The complexity of each model increased as higher-order interactions and more terms were introduced. Initially, simpler models (1 to 5) were examined for performance using standard metrics like R-squared for in-sample and out-of-sample validation.

Models 1 to 5: These models progressively introduced more interactions and terms. The goal was to assess how different combinations of predictors influenced the target variable. With each new model, the in-sample R-squared increased, but there was a risk of overfitting as more terms were added. A key part of the exploration was comparing the models' performance on both training and testing datasets to ensure they generalized well to unseen data.

Models 6 and 7: These models added even more complexity with further interactions, introducing potential challenges for model interpretability. The results suggested that Model 7 might overfit the data, making it harder to interpret compared to the simpler Model 6, which provided a more understandable relationship between the variables despite performing slightly worse on the testing data.

In summary, we examined the impact of model complexity, considering how increasing the number of predictors and interactions influenced both performance and interpretability.

https://chatgpt.com/share/6736c9ef-ac10-8006-abf4-b95ab2d591e6